rust/kernel/init.rs

```diff
@@ -0,0 +1,827 @@
+    sync::UniqueArc,
+};
+use alloc::boxed::Box;
+use core::{
+    convert::Infallible,
+    marker::{PhantomData, Unpin},
+    mem::MaybeUninit,
+    pin::Pin,
+    ptr,
+};
+
+#[doc(hidden)]
+pub mod __private;
+
+/// Initialize a type on the stack. It will be pinned:
+/// ```rust
+/// # #![allow(clippy::blacklisted_name, clippy::new_ret_no_self)]
+/// # use kernel::{init, pin_init, stack_init, pin_data, init::*};
+/// # use core::pin::Pin;
+/// pin_data! {
+///     struct Foo {
+///         a: usize,
+///         b: Bar,
+///     }
+/// }
+///
+/// pin_data! {
+///     struct Bar {
+///         x: u32,
+///     }
+/// }
+///
+/// let a = 42;
+///
+/// stack_init!(let foo = pin_init!(Foo {
+///     a,
+///     b: Bar {
+///         x: 64,
+///     },
+/// }));
+/// let foo: Result<Pin<&mut Foo>> = foo;
+/// ```
+#[macro_export]
+macro_rules! stack_init {
+    (let $var:ident $(: $t:ty)? = $val:expr) => {
+        let mut $var = $crate::init::__private::StackInit$(::<$t>)?::uninit();
+        let val = $val;
+        let mut $var = unsafe { $crate::init::__private::StackInit::init(&mut $var, val) };
+    };
+}
+
+/// Construct an in-place initializer for structs.
+///
+/// The syntax is identical to a normal struct initializer:
+/// ```rust
+/// # #![allow(clippy::blacklisted_name, clippy::new_ret_no_self)]
+/// # use kernel::{init, pin_init, pin_data, init::*};
+/// # use core::pin::Pin;
+/// pin_data! {
+///     struct Foo {
+///         a: usize,
+///         b: Bar,
+///     }
+/// }
+///
+/// pin_data! {
+///     struct Bar {
+///         x: u32,
+///     }
+/// }
+///
+/// # fn demo() -> impl PinInit<Foo> {
+/// let a = 42;
+///
+/// let initializer = pin_init!(Foo {
+///     a,
+///     b: Bar {
+///         x: 64,
+///     },
+/// });
+/// # initializer }
+/// # Box::pin_init(demo()).unwrap();
+/// ```
+/// Arbitrary rust expressions can be used to set the value of a variable.
+///
+/// # Init-functions
+///
+/// When working with this library it is often desired to let others construct your types without
+/// giving access to all fields. This is where you would normally write a plain function `new`
+/// that would return a new instance of your type. With this library that is also possible, however
+/// there are a few extra things to keep in mind.
+///
+/// To create an initializer function, simple declare it like this:
+/// ```rust
+/// # #![allow(clippy::blacklisted_name, clippy::new_ret_no_self)]
+/// # use kernel::{init, pin_init, pin_data, init::*};
+/// # use core::pin::Pin;
+/// # pin_data! { struct Foo {
+/// #     a: usize,
+/// #     b: Bar,
+/// # }}
+/// # pin_data! { struct Bar {
+/// #     x: u32,
+/// # }}
+///
+/// impl Foo {
+///     fn new() -> impl PinInit<Self> {
+///         pin_init!(Self {
+///             a: 42,
+///             b: Bar {
+///                 x: 64,
+///             },
+///         })
+///     }
+/// }
+/// ```
+/// Users of `Foo` can now create it like this:
+/// ```rust
```

Comment:
```suggestion
/// ```
///
/// Users of `Foo` can now create it like this:
///
/// ```rust
```

(and similarly elsewhere)